### PR TITLE
SRM remove multiple ENSGs per peptide

### PR DESCRIPTION
**Pull request by jaclynbeck-sage**
This PR handles the fact that multiple Ensembl IDs map to the same UniProt ID, so SRM data had duplicates with Ensembl IDs that were mostly HASGs. 

I ended up doing a bunch of cleanup across multiple notebooks as a side-effect and added it to this PR. So this PR contains the following changes:
1. **SRM processing now removes HASGs that match to peptide UniProt IDs**. I've confirmed that the only Ensembl IDs remaining for peptides that had duplicates are the ones in bold in [the JIRA ticket](https://sagebionetworks.jira.com/browse/AG-1296), and no other genes were removed from the data set. 
2. I made a new file `preprocessing_utils.py` and pulled some duplicated code into functions in that file. Right now, that's requests to Biomart and filtering HASGs, but can be expanded on if necessary. It currently lives in the same folder as the pre-processing notebooks.
3. Changed SRM pre-processing, GWAS pre-processing, and gene metadata pre-processing to use these functions instead of duplicating code. 
4. Gene metadata processing code was pretty out of date and needed some updating to work again. 
5. **Gene metadata processing now removes HASGs that are pulled in by Biomart ONLY** (it doesn't touch the ones that come from our data source files since we haven't decided how to handle that yet). The resulting `gene_table_merged.feather` file and `gene_info.json` (post-transformation) will have FEWER genes than the last update.  
6. Updating the `gene_table_merged.feather` file **pulled in Ensembl release 111**. 
7. The version numbers in both config files were changed to use the newly-generated SRM and gene table files. 

I spot-checked the new `gene_info.json` file generated by CI and it looks good to me (see also [this JIRA comment](https://sagebionetworks.jira.com/browse/AG-1296?focusedCommentId=195692) for why some HASGs are still in `gene_info`). 

The diffs on the notebooks are awful to look at so I've linked the actual notebook files below:
igap [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/6039932ec4a503fa176bfb6e366a5a72d8cfce6b/data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb)
srm [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/620c094d4e73eb37f9797f329e8009173d401269/data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb)
gene annotations [before](https://github.com/Sage-Bionetworks/agora-data-tools/blob/62fda2f739aec4c6aa1ac4c2ff638e3fa505cc39/data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb) and [after](https://github.com/Sage-Bionetworks/agora-data-tools/blob/1bf18f28b805bdd889e0188e60a72600ebafaea3/data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb)

### CODE REVIEW

File: data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb [hunk 1]
```diff
@@ -317,25 +317,20 @@
...
-    "for attr in attributes:\n",
-    "    query = query + '<Attribute name = \"' + attr + '\" />'\n",
-    "\n",
-    "query = query + '</Dataset>'\n",
-    "query = query + '</Query>'\n",
-    "    \n",
-    "response = requests.get(url = 'http://www.ensembl.org/biomart/martservice', params = {'query': query})\n",
+    "result = preprocessing_utils.manual_query_biomart(\n",
+    "    attributes=attributes, filters=filters\n",
+    ")\n",
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb:325]: Here is where I moved biomart query code to preprocessing_utils.py

File: data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb [hunk 2]
```diff
@@ -372,15 +358,10 @@
     }
    ],
    "source": [
-    "regex = re.compile(\"^[0-9|X|Y|MT]\")\n",
-    "keep = result['chromosome_name'].apply(\n",
-    "    # Keep rows if they have a numerical chromosome name, or have X, Y, or MT\n",
-    "    lambda row: re.match(regex, row) is not None if type(row) == str else True # Always true for numbers \n",
+    "result = preprocessing_utils.filter_hasgs(\n",
+    "    df=result, chromosome_name_column=\"chromosome_name\"\n",
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-796_process_gwas.ipynb:362]: Here is where I moved the HASG filter code to preprocessing_utils.py

File: data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb [hunk 3]
```diff
@@ -41,7 +41,8 @@
     "import time\n",
     "import re\n",
     "import synapseclient\n",
-    "import agoradatatools.etl.extract as extract"
+    "import agoradatatools.etl.extract as extract\n",
+    "import preprocessing_utils"
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:45]: Most of the changes in this file are from the black formatter. I'll highlight actual code changes with comments.

File: data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb [hunk 4]
```diff
@@ -442,8 +470,393 @@
...
+    ")\n",
+    "ensembl_ids[\"ENSG\"] = ensembl_ids[\"ENSG\"].str.extract(r\"(\\w+)\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eee4be20",
+   "metadata": {},
+   "source": [
+    "Filter out HASGs by querying Biomart for chromosome information and removing genes with HASG chromosomes notation. "
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:485]: NEW code starts here to filter HASGs

File: data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb [hunk 5]
```diff
@@ -442,8 +470,393 @@
...
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "ensembl_ids = ensembl_ids.loc[ensembl_ids[\"ENSG\"].isin(valid_ids[\"ENSG\"])]\n",
+    "ensembl_ids"
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-866_process_srm_diffexp_data.ipynb:860]: End new code

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 6]
```diff
@@ -101,38 +103,97 @@
...
+      "\n",
+      "[70712 rows x 3 columns]\n",
+      "70712 genes found.\n"
      ]
     }
    ],
    "source": [
     "# Sometimes Biomart doesn't respond and the command needs to be sent again. Try up to 5 times.\n",
     "for T in range(5):\n",
     "    try:\n",
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:139]: This chunk is a code update

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 7]
```diff
@@ -101,38 +103,97 @@
...
+      "63188 genes remaining after HASG filtering.\n"
+     ]
+    }
+   ],
+   "source": [
+    "ensembl_ids_df = preprocessing_utils.filter_hasgs(\n",
+    "    df=ensembl_ids_df, chromosome_name_column=\"chromosome_name\"\n",
+    ")\n",
+    "print(str(ensembl_ids_df.shape[0]) + \" genes remaining after HASG filtering.\")"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:197]: End code update

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 8]
```diff
@@ -241,73 +317,86 @@
...
-    "        ensembl_ids = ensembl_ids + file_ensembl_ids.tolist()\n",
+    "        file_ensembl_list = file_ensembl_list + file_ensembl_ids.tolist()\n",
     "        if \"n/A\" in file_ensembl_ids.tolist():\n",
     "            print(file + \" has an n/A Ensembl ID\")\n",
     "        if np.NaN in file_ensembl_ids.tolist():\n",
-    "            print(file + \" has an NaN Ensembl ID\")"
+    "            print(file + \" has an NaN Ensembl ID\")\n",
+    "    else:\n",
+    "        print(\"WARNING: no Ensembl ID column found for \" + file + \"!\")"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:357]: The couple changes above are a code update

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 9]
```diff
@@ -241,73 +317,86 @@
...
+    ")\n",
     "\n",
-    "# Convert to a pandas data frame\n",
-    "ensembl_ids_df = pd.DataFrame({'ensembl_gene_id': ensembl_ids})\n",
-    "len(ensembl_ids_df)"
+    "missing_df = pd.DataFrame({\"ensembl_gene_id\": list(missing), \"chromosome_name\": \"\"})\n",
+    "ensembl_ids_df = pd.concat([ensembl_ids_df, missing_df])\n",
+    "ensembl_ids_df = ensembl_ids_df.dropna(subset=[\"ensembl_gene_id\"])\n",
+    "print(len(ensembl_ids_df))"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:389]: The couple lines above are a code update

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 10]
```diff
@@ -710,110 +783,117 @@
...
+    "    how=\"left\",\n",
+    "    on=\"ensembl_gene_id\",\n",
+    "    validate=\"many_to_many\",\n",
+    ")\n",
     "\n",
-    "gene_table_merged = transform.standardize_column_names(gene_table_merged)\n",
-    "gene_table_merged = transform.standardize_values(gene_table_merged)\n",
+    "gene_table_merged = utils.standardize_column_names(gene_table_merged)\n",
+    "gene_table_merged = utils.standardize_values(gene_table_merged)\n",
     "\n",
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:897]: Code update above

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 11]
```diff
@@ -831,19 +911,37 @@
...
+    "            flattened = flattened + item\n",
+    "        else:\n",
+    "            flattened.append(item)\n",
+    "    return flattened\n",
+    "\n",
+    "\n",
+    "gene_table_merged[\"alias\"] = gene_table_merged[\"alias\"].apply(\n",
+    "    lambda row: list(set(flatten(row)))\n",
+    ")"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:945]: Code update above

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 12]
```diff
@@ -1782,235 +1409,88 @@
...
+       "66812         [LOC124904613, LOC124905809, LOC124905573]     NaN      NaN  "
       ]
      },
-     "execution_count": 12,
+     "execution_count": 13,
      "metadata": {},
      "output_type": "execute_result"
     }
    ],
    "source": [
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:1454]: Begin code update chunk

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 13]
```diff
@@ -1782,235 +1409,88 @@
...
     "\n",
-    "# Printed out table should have unique Ensembl IDs with aliases properly added\n",
-    "gene_table_merged.loc[gene_table_merged['ensembl_gene_id'].isin(dupe_vals['ensembl_gene_id'])]"
+    "print(\n",
+    "    str(len(all_duplicated.drop_duplicates(\"ensembl_gene_id\")))\n",
+    "    + \" duplicated genes have been processed.\"\n",
+    ")\n",
+    "gene_table_merged = keep_df.reset_index(drop=True)\n",
+    "gene_table_merged.tail(n=10)"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:1494]: End code update chunk

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 14]
```diff
@@ -3091,21 +2594,282 @@
...
+    "        \"symbol\",\n",
+    "        \"type_of_gene\",\n",
+    "        \"ensembl_release\",\n",
+    "        \"possible_replacement\",\n",
+    "        \"permalink\",\n",
+    "    ]\n",
+    "]\n",
+    "\n",
+    "gene_table_merged"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:2864]: Code update above

File: data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb [hunk 15]
```diff
@@ -3119,12 +2874,16 @@
...
    "metadata": {},
    "outputs": [],
    "source": [
-    "gene_table_merged.to_feather('../output/gene_table_merged_GRCh38.p13.feather')"
+    "gene_table_merged = gene_table_merged.sort_values(by=\"ensembl_gene_id\").reset_index(\n",
+    "    drop=True\n",
+    ")\n",
+    "gene_table_merged\n",
+    "gene_table_merged.to_feather(\"../../output/gene_table_merged_GRCh38.p14.feather\")"
    ]
```
Comment (jaclynbeck-sage) [inline data_analysis/agora/notebooks/preprocessing/AG-896_Preprocess_Gene_Annotations.ipynb:2887]: Code update above

File: data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+    result = pd.read_csv(StringIO(response.text), sep="\t")
+    return result
+
+
+def filter_hasgs(df: pd.DataFrame, chromosome_name_column: str) -> pd.DataFrame:
+    """Filters human alternative sequence genes (HASGs) from a data frame by using a regex to
+    identify them for removal. Valid genes will either have a numerical chromosome name or have
+    "X", "Y", or "MT" as the chromosome name. All other chromosome name formats correspond to
+    HASGs, pathches, and other genes that should be removed from the list.
```
Comment (JessterB) [inline data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py]: nitpick - 'pathches' looks like a typo

File: data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+    Args:
+        df (pd.DataFrame): a data frame as returned from a Biomart query, that must have a column
+                           containing chromosome names. Biomart already returns chromosome names
+                           that fit the above regex pattern, but any manual creation of chromosome
+                           names should be formatted properly so the regex works before using this
```
Comment (JessterB) [inline data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py]: curiosity - who is manually creating chromosome names, and why are they doing that?

**Review by JessterB — APPROVED** [review]
This lgtm, although I did not exhaustively review the changes to the notebooks...nor am I at all sure that I'm competent to do so. I'm wondering if there's anyone else on our team that might be capable and willing to review your more scientific/specialized code...I know Laura is interested in developing her python skills, so maybe we could convince her that it would be a good learning experience!

File: data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py (hunk 16 above)
Comment (jaclynbeck-sage) [reply data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py]: Yeah that's definitely a typo... oops. Will fix.

File: data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py (hunk 17 above)
Comment (jaclynbeck-sage) [reply data_analysis/agora/notebooks/preprocessing/preprocessing_utils.py]: Nobody is doing this. I don't know why, but I guess I added this to make the function sound 'generic'... Do you want me to just remove that part of the comment?

**Comment by jaclynbeck-sage** [comment]
I fixed the typo and also removed the comment about manually formatting chromosome names to avoid confusion. @JessterB do you want me to ask Laura to review the analysis part of the code before merging this PR?

**Comment by JessterB** [comment]
@jaclynbeck-sage Having Laura review the analysis would be great if she is willing and has the time. I don't want to block merging this for more than ~1 week though, since I'd like to get the updated SRM data ready for release.

**Comment by jaclynbeck-sage** [comment]
@JessterB I asked Laura and she said she's not at all familiar with Python. So I just described to her the steps I'm taking to process the data and she said it seemed sound. Not the same as a code review but, hopefully good enough for now?

**Comment by JessterB** [comment]
@jaclynbeck-sage Yes, that works. Thanks!

**Pull request merged**